octopus/src/main/scala/octopus/dsl.scala

```diff
@@ -58,118 +62,122 @@ object dsl {
+      macro DslMacros.ruleFieldSelectorAsync[M, T, G]
 
-    def ruleVC[V](asyncPred: V => Future[Boolean], whenInvalid: String)
-                 (implicit gen: Generic.Aux[T, V :: HNil]): AsyncValidator[T] =
+    def ruleVC[V](asyncPred: V => M[Boolean], whenInvalid: String)
+                 (implicit gen: Generic.Aux[T, V :: HNil], app: AppError[M]): AsyncValidatorM[M, T] =
       compose(AsyncValidationRules.ruleVC(asyncPred, whenInvalid))
 
-    def ruleCatchOnly[E <: Throwable : ClassTag](asyncPred: T => Future[Boolean],
+    def ruleCatchOnly[E <: Throwable : ClassTag](asyncPred: T => M[Boolean],
                                                  whenInvalid: String,
-                                                 whenCaught: E => String): AsyncValidator[T] =
+                                                 whenCaught: E => String)(implicit appError: AppError[M]): AsyncValidatorM[M, T] =
       compose(AsyncValidationRules.ruleCatchOnly(asyncPred, whenInvalid, whenCaught))
 
-    def ruleCatchNonFatal(asyncPred: T => Future[Boolean],
+    def ruleCatchNonFatal(asyncPred: T => M[Boolean],
                           whenInvalid: String,
-                          whenCaught: Throwable => String): AsyncValidator[T] =
+                          whenCaught: Throwable => String)(implicit appError: AppError[M]): AsyncValidatorM[M, T] =
       compose(AsyncValidationRules.ruleCatchNonFatal(asyncPred, whenInvalid, whenCaught))
 
-    def ruleEither(asyncPred: T => Future[Either[String, Boolean]],
-                   whenInvalid: String): AsyncValidator[T] =
+    def ruleEither(asyncPred: T => M[Either[String, Boolean]],
+                   whenInvalid: String)(implicit appError: AppError[M]): AsyncValidatorM[M, T] =
       compose(AsyncValidationRules.ruleEither(asyncPred, whenInvalid))
 
-    def ruleOption(asyncPred: T => Future[Option[Boolean]],
+    def ruleOption(asyncPred: T => M[Option[Boolean]],
                    whenInvalid: String,
-                   whenNone: String): AsyncValidator[T] =
+                   whenNone: String)(implicit appError: AppError[M]): AsyncValidatorM[M, T] =
       compose(AsyncValidationRules.ruleOption(asyncPred, whenInvalid, whenNone))
 
   }
 
-  implicit class AsyncValidatorSyncOps[T](val v: AsyncValidator[T]) extends AnyVal {
+  implicit class AsyncValidatorSyncOps[M[_], T](val v: AsyncValidatorM[M, T]) extends AnyVal {
 
-    def async: AsyncValidatorAsyncOps[T] =
-      new AsyncValidatorAsyncOps[T](v)
+    def async: AsyncValidatorAsyncOps[M, T] =
+      new AsyncValidatorAsyncOps[M, T](v)
 
-    def compose(v2: Validator[T]): AsyncValidator[T] =
-      AsyncValidator.instance { (obj: T, ec: ExecutionContext) =>
-        v.validate(obj)(ec)
-          .map(_ ++ v2.validate(obj))(ec)
+    def compose(v2: Validator[T])(implicit appError: AppError[M]): AsyncValidatorM[M, T] =
+      AsyncValidatorM.instance { (obj: T) =>
+        AppError[M].map(v.validate(obj)) {
```

Comment:
we can use implicit param directly: `appError`

Reply:
Done